server/src/test/java/org/elasticsearch/index/shard/GlobalCheckpointListenersTests.java

```diff
@@ -0,0 +1,402 @@
+            final GlobalCheckpointListeners.GlobalCheckpointListener listener =
+                    (g, e) -> {
+                        assert g == UNASSIGNED_SEQ_NO;
+                        assert e != null;
+                        if (failure) {
+                            throw new RuntimeException("failure");
+                        } else {
+                            exceptions[index] = e;
+                        }
+                    };
+            globalCheckpointListeners.add(NO_OPS_PERFORMED, listener);
+        }
+        globalCheckpointListeners.close();
+        for (int i = 0; i < numberOfListeners; i++) {
+            if (failures[i]) {
+                assertNull(exceptions[i]);
+            } else {
+                assertNotNull(exceptions[i]);
+                assertThat(exceptions[i].getShardId(), equalTo(shardId));
+            }
+        }
+        int failureCount = 0;
+        for (int i = 0; i < numberOfListeners; i++) {
+            if (failures[i]) {
+                failureCount++;
+            }
+        }
+        if (failureCount > 0) {
+            final ArgumentCaptor<String> message = ArgumentCaptor.forClass(String.class);
+            final ArgumentCaptor<RuntimeException> t = ArgumentCaptor.forClass(RuntimeException.class);
+            verify(mockLogger, times(failureCount)).warn(message.capture(), t.capture());
+            assertThat(message.getValue(), equalTo("error notifying global checkpoint listener of closed shard"));
+            assertNotNull(t.getValue());
+            assertThat(t.getValue().getMessage(), equalTo("failure"));
+        }
+    }
+
+    public void testNotificationUsesExecutor() {
+        final AtomicInteger count = new AtomicInteger();
+        final Executor executor = command -> {
+            count.incrementAndGet();
+            command.run();
+        };
+        final GlobalCheckpointListeners globalCheckpointListeners = new GlobalCheckpointListeners(shardId, executor, logger);
+        globalCheckpointListeners.globalCheckpointUpdated(NO_OPS_PERFORMED);
+        final int numberOfListeners = randomIntBetween(0, 16);
+        for (int i = 0; i < numberOfListeners; i++) {
+            globalCheckpointListeners.add(NO_OPS_PERFORMED, (g, e) -> {});
+        }
+        globalCheckpointListeners.globalCheckpointUpdated(randomLongBetween(NO_OPS_PERFORMED, Long.MAX_VALUE));
+        assertThat(count.get(), equalTo(1));
+    }
+
+    public void testConcurrency() throws BrokenBarrierException, InterruptedException {
+        final ExecutorService executor = Executors.newFixedThreadPool(randomIntBetween(1, 8));
+        final GlobalCheckpointListeners globalCheckpointListeners = new GlobalCheckpointListeners(shardId, executor, logger);
+        final AtomicLong globalCheckpoint = new AtomicLong(NO_OPS_PERFORMED);
+        globalCheckpointListeners.globalCheckpointUpdated(globalCheckpoint.get());
+        final CyclicBarrier barrier = new CyclicBarrier(3);
```

Comment:
why 3 here? How do we ensure we end up running every thread to completion? (I have a hard time counting the `barrier.await()` here)

Reply:
There are three threads. The thread repeatedly firing updates, the thread repeatedly adding listeners, and the main test thread. We use the barrier to synchronize their start (they all wait on the barrier, then go), and to signal when the updating and adding thread are done (they all wait on the barrier again). So within the updating thread we have:

```
wait on barrier
do the updating loop
wait on barrier
```

while within the adding thread we have

```
wait on barrier
do the adding loop
wait on barrier
```

and the main test thread does two waits

```
wait on barrier
wait on barrier
```

The first wait corresponds to the top wait of the updating and adding threads, and the second wait corresponds to the bottom wait of the updating and adding threads.

Does that help?

Reply:
ok, thanks for the explanation, makes sense. Can you add a comment to that effect?

Reply:
@ywelsch I pushed 07dab4ff80965bdeeb36782fa9868e028cc02981.